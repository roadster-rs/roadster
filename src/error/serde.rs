use crate::error::Error;

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum SerdeError {
    #[error(transparent)]
    Json(#[from] serde_json::Error),

    #[error(transparent)]
    TomlDeserialize(#[from] toml::de::Error),

    #[error(transparent)]
    TomlSerialize(#[from] toml::ser::Error),

    #[error(transparent)]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

impl From<serde_json::Error> for Error {
    fn from(value: serde_json::Error) -> Self {
        Self::Serde(SerdeError::from(value))
    }
}

impl From<toml::de::Error> for Error {
    fn from(value: toml::de::Error) -> Self {
        Self::Serde(SerdeError::from(value))
    }
}

impl From<toml::ser::Error> for Error {
    fn from(value: toml::ser::Error) -> Self {
        Self::Serde(SerdeError::from(value))
    }
}
