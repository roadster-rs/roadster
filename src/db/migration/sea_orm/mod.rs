use crate::app::context::AppContext;
use crate::db::migration::{DownArgs, MigrationInfo, MigrationStatus, Migrator, UpArgs};
use crate::error::RoadsterResult;
use async_trait::async_trait;
use axum_core::extract::FromRef;
use sea_orm_migration::MigratorTrait;
use std::cmp::min;
use std::marker::PhantomData;

pub mod check;
pub mod collation;
pub mod schema;
pub mod timestamp;
pub mod user;
pub mod uuid;

impl From<sea_orm_migration::Migration> for MigrationInfo {
    fn from(value: sea_orm_migration::Migration) -> Self {
        Self {
            name: value.name().to_string(),
            status: value.status().into(),
        }
    }
}

impl From<sea_orm_migration::MigrationStatus> for MigrationStatus {
    fn from(value: sea_orm_migration::MigrationStatus) -> Self {
        match value {
            sea_orm_migration::MigrationStatus::Applied => Self::Applied,
            sea_orm_migration::MigrationStatus::Pending => Self::Pending,
        }
    }
}

pub struct SeaOrmMigrator<M>
where
    M: MigratorTrait + Send + Sync,
{
    migrator: PhantomData<M>,
}

impl<M> SeaOrmMigrator<M>
where
    M: MigratorTrait + Send + Sync,
{
    pub fn new(_migrator: M) -> Self {
        Self {
            migrator: Default::default(),
        }
    }
}

#[async_trait]
impl<S, M> Migrator<S> for SeaOrmMigrator<M>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    M: MigratorTrait + Send + Sync,
{
    #[tracing::instrument(skip_all)]
    async fn up(&self, state: &S, args: &UpArgs) -> crate::error::RoadsterResult<usize> {
        let context = crate::app::context::AppContext::from_ref(state);
        let pending = M::get_pending_migrations(context.sea_orm()).await?;

        let to_run = if let Some(steps) = args.steps {
            min(steps, pending.len())
        } else {
            pending.len()
        };

        M::up(context.sea_orm(), args.steps.map(|steps| steps as u32)).await?;

        // Assume all of the pending steps (up to `args.steps` count) ran successfully.
        Ok(to_run)
    }

    #[tracing::instrument(skip_all)]
    async fn down(&self, state: &S, args: &DownArgs) -> RoadsterResult<usize> {
        let context = crate::app::context::AppContext::from_ref(state);
        let applied = M::get_applied_migrations(context.sea_orm()).await?;

        let to_roll_back = if let Some(steps) = args.steps {
            min(steps, applied.len())
        } else {
            applied.len()
        };

        M::down(context.sea_orm(), args.steps.map(|steps| steps as u32)).await?;

        // Assume all of the applied steps (up to `args.steps` count) were rolled back successfully.
        Ok(to_roll_back)
    }

    #[tracing::instrument(skip_all)]
    async fn status(&self, state: &S) -> RoadsterResult<Vec<MigrationInfo>> {
        let context = crate::app::context::AppContext::from_ref(state);

        let migrations = M::get_migration_with_status(context.sea_orm())
            .await?
            .into_iter()
            .map(|migration| migration.into())
            .collect();

        Ok(migrations)
    }
}
