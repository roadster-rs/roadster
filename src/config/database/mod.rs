use crate::util::serde::default_true;
use sea_orm::ConnectOptions;
use serde_derive::{Deserialize, Serialize};
use serde_with::serde_as;
use std::time::Duration;
use url::Url;
use validator::Validate;

#[serde_as]
#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Database {
    /// This can be overridden with an environment variable, e.g. `ROADSTER__DATABASE__URI=postgres://example:example@example:1234/example_app`
    pub uri: Url,
    /// Whether to automatically apply migrations during the app's start up. Migrations can also
    /// be manually performed via the `roadster migration [COMMAND]` CLI command.
    pub auto_migrate: bool,
    #[serde(default = "Database::default_connect_timeout")]
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub connect_timeout: Duration,
    /// Whether to attempt to connect to the DB immediately upon creating the [`ConnectOptions`].
    /// If `true` will wait to connect to the DB until the first DB query is attempted.
    #[serde(default = "default_true")]
    pub connect_lazy: bool,
    #[serde(default = "Database::default_acquire_timeout")]
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub acquire_timeout: Duration,
    #[serde_as(as = "Option<serde_with::DurationSeconds>")]
    pub idle_timeout: Option<Duration>,
    #[serde_as(as = "Option<serde_with::DurationSeconds>")]
    pub max_lifetime: Option<Duration>,
    #[serde(default)]
    pub min_connections: u32,
    pub max_connections: u32,
}

impl Database {
    fn default_connect_timeout() -> Duration {
        Duration::from_millis(1000)
    }

    fn default_acquire_timeout() -> Duration {
        Duration::from_millis(1000)
    }
}

impl From<Database> for ConnectOptions {
    fn from(database: Database) -> Self {
        ConnectOptions::from(&database)
    }
}

impl From<&Database> for ConnectOptions {
    fn from(database: &Database) -> Self {
        let mut options = ConnectOptions::new(database.uri.to_string());
        options
            .connect_timeout(database.connect_timeout)
            .connect_lazy(database.connect_lazy)
            .acquire_timeout(database.acquire_timeout)
            .min_connections(database.min_connections)
            .max_connections(database.max_connections)
            .sqlx_logging(false);
        if let Some(idle_timeout) = database.idle_timeout {
            options.idle_timeout(idle_timeout);
        }
        if let Some(max_lifetime) = database.max_lifetime {
            options.max_lifetime(max_lifetime);
        }
        options
    }
}

#[cfg(test)]
mod deserialize_tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::{assert_debug_snapshot, assert_toml_snapshot};
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        uri = "https://example.com:1234"
        auto-migrate = true
        max-connections = 1
        "#
    )]
    #[case(
        r#"
        uri = "https://example.com:1234"
        auto-migrate = true
        max-connections = 1
        connect-timeout = 1000
        acquire-timeout = 2000
        idle-timeout = 3000
        max-lifetime = 4000
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn sidekiq(_case: TestCase, #[case] config: &str) {
        let database: Database = toml::from_str(config).unwrap();

        assert_toml_snapshot!(database);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn db_config_to_connect_options() {
        let db = Database {
            uri: Url::parse("postgres://example:example@example:1234/example_app").unwrap(),
            auto_migrate: true,
            connect_timeout: Duration::from_secs(1),
            connect_lazy: true,
            acquire_timeout: Duration::from_secs(2),
            idle_timeout: Some(Duration::from_secs(3)),
            max_lifetime: Some(Duration::from_secs(4)),
            min_connections: 10,
            max_connections: 20,
        };

        let connect_options = ConnectOptions::from(&db);

        assert_debug_snapshot!(connect_options);
    }
}
