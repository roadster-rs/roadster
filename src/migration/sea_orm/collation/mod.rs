//! Utilities and migrations related to collations.

pub mod m20241022_065427_case_insensitive_collation;

use sea_orm::{DbBackend, Statement};
use sea_orm_migration::prelude::*;

/// Collations available from Roadster.
#[derive(DeriveIden)]
#[non_exhaustive]
pub enum Collation {
    /// The `default` collation. This comes included in Postgres.
    ///
    /// Note: This iden needs to be surrounded in quotes, at least in Postgres.
    Default,
    /// A case-insensitive collation.
    CaseInsensitive,
}

/// Wrapper around [`create_case_insensitive_collation`] to execute the returned [`Statement`], if
/// present.
///
/// # Examples
/// ```rust
/// use roadster::migration::sea_orm::collation::exec_create_case_insensitive_collation;
/// use sea_orm_migration::prelude::*;
///
/// #[derive(DeriveMigrationName)]
/// pub struct Migration;
///
/// #[async_trait::async_trait]
/// impl MigrationTrait for Migration {
///     async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
///         exec_create_case_insensitive_collation(manager).await
///     }
/// #
/// #    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
/// #        unimplemented!()
/// #    }
/// }
/// ```
pub async fn exec_create_case_insensitive_collation(
    manager: &SchemaManager<'_>,
) -> Result<(), DbErr> {
    let statement = create_case_insensitive_collation(manager);
    if let Some(statement) = statement {
        manager.get_connection().execute(statement).await?;
    }

    Ok(())
}

/// Create a case-insensitive collation.
///
/// See: <https://www.postgresql.org/docs/current/collation.html#COLLATION-NONDETERMINISTIC>
///
/// Note: Currently only supports Postgres. If another DB is used, will return [`None`].
pub fn create_case_insensitive_collation(manager: &SchemaManager<'_>) -> Option<Statement> {
    let backend = manager.get_database_backend();
    create_case_insensitive_collation_for_db_backend(backend)
}

fn create_case_insensitive_collation_for_db_backend(backend: DbBackend) -> Option<Statement> {
    if let DbBackend::Postgres = backend {
        Some(Statement::from_string(
            backend,
            format!(
                r#"CREATE COLLATION IF NOT EXISTS {} (
provider = icu,
locale = 'und-u-ks-level2',
deterministic = false
);
"#,
                Collation::CaseInsensitive.to_string()
            ),
        ))
    } else {
        None
    }
}

/// Wrapper around [`drop_case_insensitive_collation`] to execute the returned [`Statement`], if
/// present.
///
/// # Examples
/// ```rust
/// use roadster::migration::sea_orm::collation::exec_drop_case_insensitive_collation;
/// use sea_orm_migration::prelude::*;
///
/// #[derive(DeriveMigrationName)]
/// pub struct Migration;
///
/// #[async_trait::async_trait]
/// impl MigrationTrait for Migration {
/// #    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
/// #        unimplemented!()
/// #    }
/// #
///     async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
///         exec_drop_case_insensitive_collation(manager).await
///     }
/// }
/// ```
pub async fn exec_drop_case_insensitive_collation(
    manager: &SchemaManager<'_>,
) -> Result<(), DbErr> {
    let statement = drop_case_insensitive_collation(manager);
    if let Some(statement) = statement {
        manager.get_connection().execute(statement).await?;
    }

    Ok(())
}

/// Drop the case-insensitive collation that was previously created by [`create_case_insensitive_collation`].
///
/// Note: Currently only supports Postgres. If another DB is used, will return [None].
pub fn drop_case_insensitive_collation(manager: &SchemaManager<'_>) -> Option<Statement> {
    let backend = manager.get_database_backend();
    drop_case_insensitive_collation_for_db_backend(backend)
}

fn drop_case_insensitive_collation_for_db_backend(backend: DbBackend) -> Option<Statement> {
    if let DbBackend::Postgres = backend {
        Some(Statement::from_string(
            backend,
            format!(
                "DROP COLLATION IF EXISTS {};",
                Collation::CaseInsensitive.to_string()
            ),
        ))
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use crate::testing::snapshot::TestCase;
    use insta::assert_debug_snapshot;
    use rstest::{fixture, rstest};
    use sea_orm::DbBackend;

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(DbBackend::Postgres)]
    #[case(DbBackend::MySql)]
    #[case(DbBackend::Sqlite)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn create_case_insensitive_collation_for_db_backend(
        _case: TestCase,
        #[case] backend: DbBackend,
    ) {
        let statement = super::create_case_insensitive_collation_for_db_backend(backend);

        assert_debug_snapshot!(statement);
    }

    #[rstest]
    #[case(DbBackend::Postgres)]
    #[case(DbBackend::MySql)]
    #[case(DbBackend::Sqlite)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn drop_case_insensitive_collation_for_db_backend(_case: TestCase, #[case] backend: DbBackend) {
        let statement = super::drop_case_insensitive_collation_for_db_backend(backend);

        assert_debug_snapshot!(statement);
    }
}
