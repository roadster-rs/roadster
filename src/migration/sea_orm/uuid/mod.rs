use sea_orm::{DbBackend, Statement};

pub mod m20240726_062011_uuid_ossp_extension;

/// Create a [Statement] to create/enable the `uuid-ossp` Postgres extension
pub fn create_uuid_ossp_extension() -> Statement {
    Statement::from_string(
        DbBackend::Postgres,
        r#"CREATE EXTENSION IF NOT EXISTS "uuid-ossp";"#,
    )
}

/// Create a [Statement] to drop/disable the `uuid-ossp` Postgres extension.
pub fn drop_uuid_ossp_extension() -> Statement {
    Statement::from_string(
        DbBackend::Postgres,
        r#"DROP EXTENSION IF EXISTS "uuid-ossp";"#,
    )
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn create_uuid_ossp_extension() {
        let statement = super::create_uuid_ossp_extension();

        assert_debug_snapshot!(statement);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn drop_uuid_ossp_extension() {
        let statement = super::drop_uuid_ossp_extension();

        assert_debug_snapshot!(statement);
    }
}
