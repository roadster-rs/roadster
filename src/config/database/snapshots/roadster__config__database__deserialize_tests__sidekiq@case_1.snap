---
source: src/config/database/mod.rs
expression: database
---
uri = 'https://example.com:1234/'
auto-migrate = true
connect-timeout = 1000
connect-lazy = true
acquire-timeout = 1000
min-connections = 0
max-connections = 1
